### PR TITLE
refactor!: TrackContainer/MultiTrajectory surfaces, jacobians, meas access

### PR DESCRIPTION
**Pull request by paulgessinger**
BREAKING CHANGE: This changes the contract between `TrackContainer`, `MultiTrajectory` and their backends. Notably it

- Adds dedicated methods for surface manipulation that are used by the backend:
  - `referenceSurface_impl` to get a reference surface by index. It returns a pointer, and a new method `hasReferenceSurface` on the proxies compares it to `nullptr` to check if a reference surface is set
  - `setReferenceSurface_impl` to set a reference surface
  - `has_impl` for both backend types is not expected to respond to `referenceSurface` anymore.

- In `MultiTrajectory`, the backend is no longer expected to return an index referencing a jacobian, measurement or measurement covariance. Instead, the dedicated methods `jacobian_impl`, `measurement_impl` and `measurementCovariance_impl` are now called by the proxy with the state index directly, leaving the backend to implement index traversal. `has_impl` is not expected to respond to `calibrated`, `calibratedCov`, `jacobian` anymore.

### CODE REVIEW

File: Core/include/Acts/EventData/MultiTrajectory.hpp [hunk 1]
```diff
@@ -471,7 +471,9 @@ class TrackStateProxy {
     typeFlags() = other.typeFlags();
 
     // can be nullptr, but we just take that
```
Comment (EdwardMoyse) [inline Core/include/Acts/EventData/MultiTrajectory.hpp:473]: I guess this comment isn't really valid any more?

**Review by EdwardMoyse — COMMENTED** [review]
Looks good. I had one very minor comment. 

So, for ATLAS we need to have a matching MR for `MultiTrajectory.h` when we update to this ACTS version?

File: Core/include/Acts/EventData/MultiTrajectory.hpp (hunk 1 above)
Comment (paulgessinger) [reply Core/include/Acts/EventData/MultiTrajectory.hpp:473]: Good catch, thanks!

**Comment by paulgessinger** [comment]
> Looks good. I had one very minor comment.
> 
> So, for ATLAS we need to have a matching MR for `MultiTrajectory.h` when we update to this ACTS version?

Yes exactly. Pragmatically, when this PR is merged, my canary CI build of ACTS main in Athena is going to fail (https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines), and we (probably: I) will add the corresponding fixes to the [canary branch](https://gitlab.cern.ch/acts/athena/-/tree/canary)

**Comment by paulgessinger** [comment]
Could you approve this @EdwardMoyse ?

**Review by EdwardMoyse — APPROVED** [review]

**Pull request merged**